900 - AdopterFoster: Add adopted pets

sarvaiyanidhi opened the pull request:
# 🔗 Issue
#900 

# ✍️ Description
Displaying Adopted Pets for Adopter account and listing its files on click of each pet.

# 📷 Screenshots/Demos

![Alta-Pet-Rescue-Adopted-Pets-](https://github.com/user-attachments/assets/e39fb416-1914-44c5-bc23-87b02a2e4b82)

kasugaijin commented:
@sarvaiyanidhi thanks for doing this, one of us will review this soon.

kasugaijin commented on app/views/organizations/adopter_fosterer/adopted_pets/_pet_files.html.erb:
```diff
@@ -0,0 +1,36 @@
+<div id="pet_files">  
```
This table is almost identical to the one we use on staff/pets to upload and show files, right? `app/views/organizations/staff/pets/_file_attachment_table.html.erb` The main difference is whether or not we want to render a delete button.

I am wondering if instead of duplicating a lof ot this logic here, if we can move this partial `app/views/organizations/staff/pets/_file_attachment_table.html.erb` to a shared folder e.g. at `aap/views/organizations/shared` and then conditionally render the delete button based on the User type, e.g. if the user role is staff we render the delete button. Does that make sense?

sarvaiyanidhi replied:
Thanks for the input. Common file does make sense and easy to maintain. Have updated changes for this.

kasugaijin commented on app/controllers/organizations/adopter_fosterer/adopted_pets_controller.rb:
```diff
@@ -0,0 +1,33 @@
+        .where(status: :adoption_made, profile_show: true), with: Organizations::AdopterFosterAdoptedPetPolicy)
+      end
+
+      def files
```
We could create a new route and controller to handle this instead e.g., 
```
module Organizations
  module AdopterFosterer
    class AdoptedPetFilesController
```

And use the index action to get the list of files for a given pet. This will keep things RESTful so we don't have to use custom methods. What do you think?

sarvaiyanidhi replied:
Nice.. I liked this approach to keep things more RESTful. Have made the required changes and created new files_controller.rb

kasugaijin commented on app/helpers/organizations/adopter_fosterer/adopted_pets_helper.rb:
i think we can get rid of this? I assume this was create when using a Rails generator?

kasugaijin commented on app/javascript/controllers/card_controller.js:
I like this, it can be re-used for other situations with multiple cards to select from.

kasugaijin commented on app/controllers/organizations/adopter_fosterer/adopted_pets_controller.rb:
```diff
@@ -0,0 +1,33 @@
+      before_action :context_authorize!
+      layout "adopter_foster_dashboard"
+
+      def index
```
Could we instead query the `Match` model to find all matches with this `adopter_foster_account_id` where the `match_type: :adoption`?

kasugaijin commented on app/policies/organizations/adopter_foster_adopted_pet_policy.rb:
```diff
@@ -0,0 +1,15 @@
+class Organizations::AdopterFosterAdoptedPetPolicy < ApplicationPolicy
```
Ah yes this is close to what I had in mind with this comment https://github.com/rubyforgood/pet-rescue/pull/910/files#r1720837255...I should have looked further down :) 

Still, we should be querying Match instead of AdopterApplication. 
Also, this name `Organizations::AdopterFosterAdoptedPetPolicy` is duplicating a common module name in this app `AdopterFosterer` so let's add that directory so we have `Organizations::AdopterFosterer::MatchPolicy` (I think the policy names ideally mirror the model names, right?)

kasugaijin commented on config/routes.rb:
```diff
@@ -61,6 +61,11 @@
       resources :dashboard, only: [:index]
       resources :likes, only: [:index, :create, :destroy]
       resources :adopter_applications, path: "applications", only: %i[index create update]
+      resources :adopted_pets, only: [:index] do
```
https://github.com/rubyforgood/pet-rescue/pull/910/files#r1720837255 to follow up on this comment.

I think something like below is cleaner. 
```
resources :adopted_pets, only: [:index] do
  resources :files, only: [:index], module: :adopted_pets
end
```

kasugaijin commented on test/integration/organizations/adopter_fosterer/adopted_pet_test.rb:
I think controller tests are more maintinable in the long run. If we start asserting dom elements it makes the tests very fragile when we come to modify the UI, which happens often enough to become annoying :)

I think we should keep the policy tests, and add controller tests to assert the actions return what we expect in the collections, and then get rid of integration tests asserting DOM elements and System tests.

kasugaijin reviewed (commented):
@sarvaiyanidhi great start on this! The functionality is all good. I have some comments about structure and approach in some cases. Let me know if you have any questions on these comments :)

sarvaiyanidhi commented:
> @sarvaiyanidhi great start on this! The functionality is all good. I have some comments about structure and approach in some cases. Let me know if you have any questions on these comments :)

Thanks @kasugaijin for the detailed feedback. Have gone through the comments and will push the required changes soon.

kasugaijin commented:
@sarvaiyanidhi how's this going?

sarvaiyanidhi commented:
> @sarvaiyanidhi how's this going?

Hey sorry for the delay... Will work and push changes in 1-2 days

kasugaijin commented:
@sarvaiyanidhi not at all! No rush. Just making sure you’re not blocked.

sarvaiyanidhi commented:
Hi @kasugaijin,

Sorry took some more time than expected as was tied up with few things.. I have addressed all points and pushed code. Can you please check if changes made for query using Match model is as per requirement. Also, I am facing one issue with authorize test in controller test, not able to figure why it is failing if you can help with it.

Thanks

kasugaijin commented:
> Hi @kasugaijin,
> 
> Sorry took some more time than expected as was tied up with few things.. I have addressed all points and pushed code. Can you please check if changes made for query using Match model is as per requirement. Also, I am facing one issue with authorize test in controller test, not able to figure why it is failing if you can help with it.
> 
> Thanks

Thanks @sarvaiyanidhi. Can you post a picture of the error screenshot in the pet rescue slack channel? It's the fastest way to get unblocked.

kasugaijin reviewed (approved):
Awesome @sarvaiyanidhi !
